sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -590,7 +591,12 @@ object ScalaReflection extends ScalaReflection {
               "cannot be used as field name\n" + walkedTypePath.mkString("\n"))
           }
 
-          val fieldValue = Invoke(inputObject, fieldName, dataTypeFor(fieldType))
+          // we know primitive type takes only non-null, or
+          // we can infer correct nullability for struct's fieldValue by a guard using If(IsNull())
```

Comment:
Once #15979 is merged, we can use AssertNotNull approach in above discussion. This comment can be simplified then to remove the second case.